controllers/database_controller.go

```diff
@@ -0,0 +1,282 @@
+
+				err = r.Create(ctx, secret)
+				if err != nil {
+					r.Log.Error(err, fmt.Sprintf("error creating Secret %s/%s", req.Namespace, secretName))
+					break
+				} else {
+					r.Log.Info("Created credential secret", "db", req.Name)
+				}
+			}
+		}
+
+		readName := fmt.Sprintf("%s-read", req.Name)
+		writeName := fmt.Sprintf("%s-write", req.Name)
+
+		err = r.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: secretName}, secret)
+		if err != nil {
+			r.Log.Error(err, "error getting secret", "secret", secretName)
+			break
+		}
+
+		users := []User{{name: readName, password: string(secret.Data[readName])}, {name: writeName, password: string(secret.Data[writeName])}}
+
+		err = r.createUsers(ctx, apiClient, users)
+		if err != nil {
+			r.Log.Error(err, "error creating users", "users", users)
+			break
+		}
+
+		roles := []string{readName, writeName}
+		err = r.createRoles(ctx, apiClient, roles)
+		if err != nil {
+			r.Log.Error(err, "error creating roles", "roles", roles)
+			break
+		}
+
+		readPermissions := []stardogapi.Permission{
+			{ResourceType: "db", Action: "READ", Resources: []string{req.Name}},
+			{ResourceType: "metadata", Action: "READ", Resources: []string{req.Name}},
+		}
+
+		writePermissions := []stardogapi.Permission{
+			{ResourceType: "db", Action: "WRITE", Resources: []string{req.Name}},
+			{ResourceType: "metadata", Action: "WRITE", Resources: []string{req.Name}},
+		}
+
+		// TODO remove unwanted permissions?
+		err = r.addPermissions(ctx, apiClient, readName, readPermissions)
+		if err != nil {
+			r.Log.Error(err, "adding permission to role failed", "role", readName, "permission", readPermissions)
+			break
+		}
+		err = r.addPermissions(ctx, apiClient, writeName, writePermissions)
+		if err != nil {
+			r.Log.Error(err, "adding permission to role failed", "role", writeName, "permission", writePermissions)
+			break
+		}
+
+		// assign roles
+		readUserRoles, err := apiClient.GetUserRoles(ctx, readName)
+		if err != nil {
+			r.Log.Error(err, "error getting user roles", "user", readName)
+			break
+		}
+
+		if len(readUserRoles) != 1 || readUserRoles[0] != readName {
+			err = apiClient.SetUserRoles(ctx, readName, []string{readName})
+			if err != nil {
+				r.Log.Error(err, "error assigning roles to user", "user", readName, "roles", readUserRoles)
+				break
+			}
+		}
+
+		// TODO check if we actually need read when assigning write
+		writeUserRoles, err := apiClient.GetUserRoles(ctx, writeName)
+		if err != nil {
+			r.Log.Error(err, "error getting user roles", "user", writeName)
+			break
+		}
+
+		if len(writeUserRoles) != 2 || writeUserRoles[0] != readName || writeUserRoles[1] != writeName {
+			err = apiClient.SetUserRoles(ctx, writeName, []string{readName, writeName})
+			if err != nil {
+				r.Log.Error(err, "error assigning roles to user", "user", readName, "roles", readUserRoles)
+				break
+			}
+		}
+	}
+
+	return ctrl.Result{}, err
```

Comment:
This won't return any errors if an error condition happens and leads to a loop termination. In consequence the controller thinks all is well and won't try again unless restarted.

If the loop is really needed the inner part might be moved to a function to be able to improve error handling.